reconcile/dynatrace_token_provider.py

```diff
@@ -0,0 +1,404 @@
+                    if token_name == DYNATRACE_INGESTION_TOKEN_NAME:
+                        if not dry_run:
+                            ingestion_token = self.create_dynatrace_ingestion_token(
+                                dt_client, cluster.ocm_cluster.external_id
+                            )
+                            token["id"] = ingestion_token.id
+                            token["token"] = ingestion_token.token
+                        logging.info(
+                            f"Ingestion token created in Dynatrace for cluster {cluster.ocm_cluster.external_id}."
+                        )
+                    elif token_name == DYNATRACE_OPERATOR_TOKEN_NAME:
+                        if not dry_run:
+                            operator_token = self.create_dynatrace_operator_token(
+                                dt_client, cluster.ocm_cluster.external_id
+                            )
+                            token["id"] = operator_token.id
+                            token["token"] = operator_token.token
+                        logging.info(
+                            f"Operator token created in Dynatrace for cluster {cluster.ocm_cluster.external_id}."
+                        )
+                else:
+                    if token_name == DYNATRACE_INGESTION_TOKEN_NAME:
+                        ingestion_token = ApiTokenCreated(raw_element=token)
+                    elif token_name == DYNATRACE_OPERATOR_TOKEN_NAME:
+                        operator_token = ApiTokenCreated(raw_element=token)
+            if need_patching:
+                patch_syncset_payload = self.construct_base_syncset(
+                    ingestion_token=ingestion_token, operator_token=operator_token
+                )
+                try:
+                    patch_syncset(
+                        ocm_client,
+                        cluster_id=cluster.ocm_cluster.id,
+                        syncset_id=SYNCSET_ID,
+                        syncset_map=patch_syncset_payload,
+                    )
+                    logging.info("Successfully patched syncset.")
+                except Exception as e:
+                    _expose_errors_as_service_log(
+                        ocm_client,
+                        cluster.ocm_cluster.external_id,
+                        f"DTP can't patch Syncset {SYNCSET_ID} due to {str(e.args)}",
+                    )
+
+    def get_syncset(
+        self, ocm_client: OCMBaseClient, cluster: ClusterDetails
+    ) -> Mapping:
+        try:
+            syncset = get_syncset(ocm_client, cluster.ocm_cluster.id, SYNCSET_ID)
+        except Exception as e:
+            if "Not Found" in e.args[0]:
+                syncset = None
+            else:
+                raise e
+        return syncset
+
+    def token_exist_in_dynatrace(self, dt_client: Dynatrace, token_id: str) -> bool:
+        try:
+            result = dt_client.tokens.get(token_id)
```

Comment:
is there a way to fetch all tokens with one API call?

Reply:
For both ingestion and operator token? The only GET APIs are list all token or get a specific token: https://www.dynatrace.com/support/help/dynatrace-api/environment-api/tokens-v2/api-tokens

Reply:
Ah I see. We'd still need to make one API call to each dynatrace environment, but potentially we can get all the tokens from that environment and filter through by prefix finding the relevant ones

Reply:
awesome!